### ISSUE TITLE
Use custom jinja2 templates

### ISSUE THREAD

**Issue by paulocoutinhox** [pydjinni/pydjinni#14]
Hi,

There is any option to use custom jinja2 templates instead of oficial templates?

Thanks.

**Comment by jothepro** [comment]
No, this is not supported atm. You cannot override/customize the jinja2 templates that are used to render the output. 
Currently the only way to customize the output would be to write a custom generator plugin, e.g. by copying the `java` generator into a separate plugin and changing the templates there.
Providing an example for an external generator plugin is on my todo.

I'm open for discussion on how to elegantly provide support for "patching" jinja2 templates though, I do see the appeal! Maybe one could do sth similar to mkdocs [`custom_dir`](https://www.mkdocs.org/user-guide/customizing-your-theme/#using-the-theme-custom_dir)? 🤔

**Comment by paulocoutinhox** [comment]
Yes, we can have a "custom-dir" parameter and, if it exists, the model can be taken first from the custom dir and then from the official directory, if it does not exist in custom dir.

**Comment by mutagene** [comment]
Neat feature! Seems really quite powerful. To be really flexible, it seems to me that the generator should (e.g. for c++) attempt to populate both `.cpp` and `.h` files for each type (and then removing the output if it is empty?).

**Comment by mutagene** [comment]
I've been thinking a bit more, and I think it could be nice to have the ability to not just specify alternate templates, but also to specify additional custom templates. In that case, adding something like e.g. json serialisation could be done by specifying an additional set of templates for this feature, which the cpp generator would use to generate "{djinni_type}+json.cpp" and "{djinni_type}+json.hpp" as appropriate. Because they would be generated in addition to the base set of templates, users wouldn't have to worry as much about updating these templates when the base templates got updated. For something like json serialisation/deserialisation, this would also allow users to implement the feature themselves using the library of their choice without the core djinni requiring all kind of customisation and corresponding test cases.
